File: AnkiDroid/src/main/java/com/ichi2/anki/Previewer.java
/***************************************************************************************
 * Copyright (c) 2011 Kostas Spyropoulos <inigo.aldana@gmail.com>                       *
 * Copyright (c) 2013 Jolta Technologies                                                *
 * Copyright (c) 2014 Bruno Romero de Azevedo <brunodea@inf.ufsm.br>                    *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/

package com.ichi2.anki;

import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.ichi2.anki.cardviewer.PreviewLayout;
import com.ichi2.anki.cardviewer.ViewerCommand;
import com.ichi2.libanki.Card;
import com.ichi2.libanki.Collection;
import com.ichi2.libanki.Note;
import com.ichi2.libanki.Utils;

import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.LongStream;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;
import timber.log.Timber;

import static com.ichi2.anim.ActivityTransitionAnimation.Direction.END;
import static com.ichi2.anim.ActivityTransitionAnimation.Direction.START;

/**
 * The previewer intent must supply an array of cards to show and the index in the list from where
 * to begin showing them. Special rules are applied if the list size is 1 (i.e., no scrolling
 * buttons will be shown).
 */
public class Previewer extends AbstractFlashcardViewer {
    private long[] mCardList;
    private int mIndex;
    private boolean mShowingAnswer;
    private SeekBar mProgressSeekBar;
    private TextView mProgressText;

    /** Communication with Browser */
    private boolean mReloadRequired;
    private boolean mNoteChanged;

    protected PreviewLayout mPreviewLayout;

    @CheckResult
    @NonNull
    public static Intent getPreviewIntent(Context context, int index, long[] cardList) {
        Intent intent = new Intent(context, Previewer.class);
        intent.putExtra("index", index);
        intent.putExtra("cardList", cardList);
        return intent;
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        if (showedActivityFailedScreen(savedInstanceState)) {
            return;
        }
        Timber.d("onCreate()");
        super.onCreate(savedInstanceState);

        mCardList = getIntent().getLongArrayExtra("cardList");
        mIndex = getIntent().getIntExtra("index", -1);

        if (savedInstanceState != null) {
            mIndex = savedInstanceState.getInt("index", mIndex);
            mShowingAnswer = savedInstanceState.getBoolean("showingAnswer", mShowingAnswer);
            mReloadRequired = savedInstanceState.getBoolean("reloadRequired");
            mNoteChanged = savedInstanceState.getBoolean("noteChanged");
        }

        if (mCardList.length == 0 || mIndex < 0 || mIndex > mCardList.length - 1) {
            Timber.e("Previewer started with empty card list or invalid index");
            finishWithoutAnimation();
            return;
        }
        showBackIcon();
        // Ensure navigation drawer can't be opened. Various actions in the drawer cause crashes.
        disableDrawerSwipe();
        startLoadingCollection();
        initPreviewProgress();
    }

    private void initPreviewProgress() {
        mProgressSeekBar = findViewById(R.id.preview_progress_seek_bar);
        mProgressText = findViewById(R.id.preview_progress_text);
        LinearLayout progressLayout = findViewById(R.id.preview_progress_layout);

        //Show layout only when the cardList is bigger than 1
        if (mCardList.length > 1) {
            progressLayout.setVisibility(View.VISIBLE);
            mProgressSeekBar.setMax(mCardList.length - 1);
            setSeekBarListener();
            updateProgress();
        }
    }

    private void setSeekBarListener() {
        mProgressSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    mIndex = progress;
                    updateProgress();
                    setCurrentCard(getCol().getCard(mCardList[mIndex]));
                    displayCardQuestion();

                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                // Mandatory override, but unused
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (mIndex >= 0 && mIndex < mCardList.length) {
                    setCurrentCard(getCol().getCard(mCardList[mIndex]));
                    displayCardQuestion();
                }
            }
        });
    }

    @Override
    protected void onCollectionLoaded(Collection col) {
        super.onCollectionLoaded(col);
        setCurrentCard(col.getCard(mCardList[mIndex]));

        displayCardQuestion();
        if (mShowingAnswer) {
            displayCardAnswer();
        }

        showBackIcon();
    }

    /** Given a new collection of card Ids, find the 'best' valid card given the current collection
     * We define the best as searching to the left, then searching to the right of the current element
     * This occurs as many cards can be deleted when editing a note (from the Card Template Editor) */
    private int getNextIndex(List<Long> newCardList) {
        HashSet<Long> validIndices = new HashSet<>(newCardList);

        for (int i = mIndex; i >= 0; i--) {
            if (validIndices.contains(mCardList[i])) {
                return newCardList.indexOf(mCardList[i]);
            }
        }

        for (int i = mIndex + 1; i < validIndices.size(); i++) {
            if (validIndices.contains(mCardList[i])) {
                return newCardList.indexOf(mCardList[i]);
            }
        }

        throw new IllegalStateException("newCardList was empty");
    }


    @Override
    protected void setTitle() {
        getSupportActionBar().setTitle(R.string.preview_title);
    }


    @Override
    protected void initLayout() {
        super.initLayout();
        mTopBarLayout.setVisibility(View.GONE);

        findViewById(R.id.answer_options_layout).setVisibility(View.GONE);

        mPreviewLayout = PreviewLayout.createAndDisplay(this, mToggleAnswerHandler);
        mPreviewLayout.setOnNextCard((view) -> changePreviewedCard(true));
        mPreviewLayout.setOnPreviousCard((view) -> changePreviewedCard(false));
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_edit) {
            editCard();
            return true;
        } else if (item.getItemId() == R.id.action_delete) {
            showDeleteNoteDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    protected void showDeleteNoteDialog() {
        Resources res = getResources();
        new MaterialDialog.Builder(this)
                .title(res.getString(R.string.delete_card_title))
                .iconAttr(R.attr.dialogErrorIcon)
                .content(res.getString(R.string.delete_note_message,
                        Utils.stripHTML(mCurrentCard.q(true))))
                .positiveText(R.string.dialog_positive_delete)
                .negativeText(R.string.dialog_cancel)
                .onPositive((dialog, which) -> {
                    Timber.i("AbstractFlashcardViewer:: OK button pressed to delete note %d", mCurrentCard.getNid());
                    mSoundPlayer.stopSounds();
                    deleteNoteWithoutConfirmation();
                })
                .build().show();
    }


    @Override
    protected void deleteNoteWithoutConfirmation() {
        //create an ArrayList from the Array
        List<Long> tempList = LongStream.of(mCardList)
                                .boxed()
                                .collect(Collectors.toList());

        Note currentNote = mCurrentCard.getCol().getNote(mCurrentCard.getNid());

        //remove all the cards under the note from the list
        for(Card card: currentNote.cards()) {
            tempList.remove(card.getId());
        }

        //TODO: needs to be more accurate
        //decide which index to change to after deleting depending on current index
        mIndex -= (mIndex != mCardList.length - 1) ? currentNote.numberOfCards() : currentNote.numberOfCards() - 1;

        //handling case of deleting multiple cards from first index, so next index is first as well
        if(mIndex < -1)
        {
            mIndex = -1;
        }

        //converting back to Array
        mCardList = tempList.stream()
                .mapToLong(Long::longValue)
                .toArray();

        //deleting the note
        super.deleteNoteWithoutConfirmation();

        //if no cards remain, go back to the Deck Picker
        if(mCardList.length == 0) {
            openDeckPicker();
            return;
        }

        //if only 1 card is present, don't show progressLayout
        if(mCardList.length == 1) {
            LinearLayout progressLayout = findViewById(R.id.preview_progress_layout);
            progressLayout.setVisibility(View.GONE);
        } else {
            mProgressSeekBar.setMax(mCardList.length - 1);
        }

        changePreviewedCard(mIndex < mCardList.length-1);
    }


    void openDeckPicker() {
        Timber.i("Navigating to decks");
        Intent deckPicker = new Intent(Previewer.this, DeckPicker.class);
        // opening DeckPicker should use the instance on the back stack & clear back history
        deckPicker.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        startActivityWithAnimation(deckPicker, END);
    }


    @Override
    public void onBackPressed() {
        setResult(RESULT_OK, getResultIntent());
        openCardBrowser();
    }


    @Override
    protected void onNavigationPressed() {
        setResult(RESULT_OK, getResultIntent());
        openCardBrowser();
    }


    //Need to override and open, otherwise the deleted cards are not updated
    //in the card browser when navigated back to it
    @Override
    protected void openCardBrowser() {
        Intent intent = new Intent(Previewer.this, CardBrowser.class);
        Long currentCardId = getCurrentCardId();
        if (currentCardId != null) {
            intent.putExtra("currentCard", currentCardId);
        }
        startActivityForResultWithAnimation(intent, REQUEST_BROWSE_CARDS, END);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.previewer, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putLongArray("cardList", mCardList);
        outState.putInt("index", mIndex);
        outState.putBoolean("showingAnswer", mShowingAnswer);
        outState.putBoolean("reloadRequired", mReloadRequired);
        outState.putBoolean("noteChanged", mNoteChanged);
        super.onSaveInstanceState(outState);
    }


    @Override
    public void displayCardQuestion() {
        super.displayCardQuestion();
        mShowingAnswer = false;
        updateButtonsState();
    }


    // Called via mFlipCardListener in parent class when answer button pressed
    @Override
    protected void displayCardAnswer() {
        super.displayCardAnswer();
        mShowingAnswer = true;
        updateButtonsState();
    }


    @Override
    protected void hideEaseButtons() {
        /* do nothing */
    }

    @Override
    protected void displayAnswerBottomBar() {
        /* do nothing */
    }


    @Override
    public boolean executeCommand(@NonNull ViewerCommand which) {
        /* do nothing */
        return false;
    }


    @Override
    protected void performReload() {
        mReloadRequired = true;
        List<Long> newCardList = getCol().filterToValidCards(mCardList);

        if (newCardList.isEmpty()) {
            finishWithoutAnimation();
            return;
        }

        mIndex = getNextIndex(newCardList);
        mCardList = Utils.collection2Array(newCardList);
        setCurrentCard(getCol().getCard(mCardList[mIndex]));
        displayCardQuestion();
    }


    @Override
    protected void onEditedNoteChanged() {
        super.onEditedNoteChanged();
        mNoteChanged = true;
    }


    protected void changePreviewedCard(boolean nextCard) {
        mIndex = nextCard ? mIndex + 1 : mIndex - 1;
        setCurrentCard(getCol().getCard(mCardList[mIndex]));
        displayCardQuestion();
        updateProgress();
    }


    private final View.OnClickListener mToggleAnswerHandler = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            if (mShowingAnswer) {
                displayCardQuestion();
            } else {
                displayCardAnswer();
            }
        }
    };

    private void updateButtonsState() {
        mPreviewLayout.setShowingAnswer(mShowingAnswer);

        // If we are in single-card mode, we show the "Show Answer" button on the question side
        // and hide navigation buttons.
        if (mCardList.length == 1) {
            mPreviewLayout.hideNavigationButtons();
            return;
        }

        mPreviewLayout.setPrevButtonEnabled(mIndex > 0);
        mPreviewLayout.setNextButtonEnabled(mIndex < mCardList.length - 1);
    }

    private void updateProgress() {
        if (mProgressSeekBar.getProgress() != mIndex) {
            mProgressSeekBar.setProgress(mIndex);
        }

        String progress = getString(R.string.preview_progress_bar_text, mIndex + 1, mCardList.length);
        mProgressText.setText(progress);
    }

    @NonNull
    private Intent getResultIntent() {
        Intent intent = new Intent();
        intent.putExtra("reloadRequired", mReloadRequired);
        intent.putExtra("noteChanged", mNoteChanged);
        return intent;
    }
}

File: AnkiDroid/src/main/java/com/ichi2/anki/Previewer.kt
/***************************************************************************************
 * Copyright (c) 2011 Kostas Spyropoulos <inigo.aldana@gmail.com>                       *
 * Copyright (c) 2013 Jolta Technologies                                                *
 * Copyright (c) 2014 Bruno Romero de Azevedo <brunodea@inf.ufsm.br>                    *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/
package com.ichi2.anki

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.annotation.CheckResult
import androidx.annotation.VisibleForTesting
import com.ichi2.anki.cardviewer.Gesture
import com.ichi2.anki.cardviewer.PreviewLayout
import com.ichi2.anki.cardviewer.ViewerCommand
import com.ichi2.annotations.NeedsTest
import com.ichi2.libanki.Collection
import com.ichi2.utils.performClickIfEnabled
import timber.log.Timber

/**
 * The previewer intent must supply an array of cards to show and the index in the list from where
 * to begin showing them. Special rules are applied if the list size is 1 (i.e., no scrolling
 * buttons will be shown).
 */

class Previewer : AbstractFlashcardViewer() {
    private lateinit var cardList: LongArray
    private var index = 0
    private var showingAnswer = false
    private lateinit var progressSeekBar: SeekBar
    private lateinit var progressText: TextView

    /** Communication with Browser  */
    private var reloadRequired = false
    private var noteChanged = false
    private var previewLayout: PreviewLayout? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        if (showedActivityFailedScreen(savedInstanceState)) {
            return
        }
        super.onCreate(savedInstanceState)
        cardList = requireNotNull(intent.getLongArrayExtra("cardList")) { "'cardList' required" }
        index = intent.getIntExtra("index", -1)
        if (savedInstanceState != null) {
            index = savedInstanceState.getInt("index", index)
            showingAnswer = savedInstanceState.getBoolean("showingAnswer", showingAnswer)
            reloadRequired = savedInstanceState.getBoolean("reloadRequired")
            noteChanged = savedInstanceState.getBoolean("noteChanged")
        }
        if (cardList.isEmpty() || index < 0 || index > cardList.size - 1) {
            Timber.e("Previewer started with empty card list or invalid index")
            finish()
            return
        }
        showBackIcon()
        // Ensure navigation drawer can't be opened. Various actions in the drawer cause crashes.
        disableDrawerSwipe()
        startLoadingCollection()
        initPreviewProgress()
        setOkResult()
    }

    private fun initPreviewProgress() {
        progressSeekBar = findViewById(R.id.preview_progress_seek_bar)
        progressText = findViewById(R.id.preview_progress_text)
        val progressLayout = findViewById<LinearLayout>(R.id.preview_progress_layout)

        // Show layout only when the cardList is bigger than 1
        if (cardList.size > 1) {
            progressLayout.visibility = View.VISIBLE
            progressSeekBar.max = cardList.size - 1
            setSeekBarListener()
            updateProgress()
        }
    }

    private fun setSeekBarListener() {
        progressSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    index = progress
                    updateProgress()
                    currentCard = getColUnsafe.getCard(cardList[index])
                    displayCardQuestion()
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                // Mandatory override, but unused
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                if (index >= 0 && index < cardList.size) {
                    currentCard = getColUnsafe.getCard(cardList[index])
                    displayCardQuestion()
                }
            }
        })
    }

    override fun onCollectionLoaded(col: Collection) {
        super.onCollectionLoaded(col)
        currentCard = col.getCard(cardList[index])
        displayCardQuestion()
        if (showingAnswer) {
            displayCardAnswer()
        }
        showBackIcon()
    }

    /** Given a new collection of card Ids, find the 'best' valid card given the current collection
     * We define the best as searching to the left, then searching to the right of the current element
     * This occurs as many cards can be deleted when editing a note (from the Card Template Editor)  */
    private fun getNextIndex(newCardList: List<Long>): Int {
        val validIndices = HashSet(newCardList)
        for (i in index downTo 0) {
            if (validIndices.contains(cardList[i])) {
                return newCardList.indexOf(cardList[i])
            }
        }
        for (i in index + 1 until validIndices.size) {
            if (validIndices.contains(cardList[i])) {
                return newCardList.indexOf(cardList[i])
            }
        }
        throw IllegalStateException("newCardList was empty")
    }

    override fun initLayout() {
        super.initLayout()
        topBarLayout!!.visibility = View.GONE
        findViewById<View>(R.id.answer_options_layout).visibility = View.GONE
        findViewById<View>(R.id.bottom_area_layout).visibility = View.VISIBLE
        previewLayout = PreviewLayout.createAndDisplay(this, toggleAnswerHandler)
        previewLayout!!.setOnNextCard { changePreviewedCard(true) }
        previewLayout!!.setOnPreviousCard { changePreviewedCard(false) }
    }

    @NeedsTest("handling: when valid and invalid")
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_DPAD_LEFT -> {
                Timber.i("Left pressed: previous card")
                previewLayout?.prevCard?.performClickIfEnabled()
                return true
            }
            KeyEvent.KEYCODE_DPAD_RIGHT -> {
                Timber.i("Right pressed: next card")
                previewLayout?.nextCard?.performClickIfEnabled()
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_edit) {
            editCard()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onNavigationPressed() {
        setOkResult()
        super.onNavigationPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.previewer, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putLongArray("cardList", cardList)
        outState.putInt("index", index)
        outState.putBoolean("showingAnswer", showingAnswer)
        outState.putBoolean("reloadRequired", reloadRequired)
        outState.putBoolean("noteChanged", noteChanged)
        super.onSaveInstanceState(outState)
    }

    override fun displayCardQuestion() {
        super.displayCardQuestion()
        showingAnswer = false
        updateButtonsState()
    }

    // Called via mFlipCardListener in parent class when answer button pressed
    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
    override fun displayCardAnswer() {
        super.displayCardAnswer()
        showingAnswer = true
        updateButtonsState()
    }

    override fun hideEaseButtons() {
        /* do nothing */
    }

    override fun displayAnswerBottomBar() {
        /* do nothing */
    }

    override fun executeCommand(which: ViewerCommand, fromGesture: Gesture?): Boolean {
        /* do nothing */
        return false
    }

    override fun performReload() {
        reloadRequired = true
        setOkResult()
        val newCardList = getColUnsafe.filterToValidCards(cardList)
        if (newCardList.isEmpty()) {
            finish()
            return
        }
        index = getNextIndex(newCardList)
        cardList = newCardList.toLongArray()
        currentCard = getColUnsafe.getCard(cardList[index])
        displayCardQuestion()
    }

    override fun onEditedNoteChanged() {
        super.onEditedNoteChanged()
        noteChanged = true
        setOkResult()
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    internal fun changePreviewedCard(nextCard: Boolean) {
        index = if (nextCard) index + 1 else index - 1
        currentCard = getColUnsafe.getCard(cardList[index])
        displayCardQuestion()
        updateProgress()
    }

    private val toggleAnswerHandler = View.OnClickListener {
        if (showingAnswer) {
            displayCardQuestion()
        } else {
            displayCardAnswer()
        }
    }

    private fun updateButtonsState() {
        previewLayout!!.setShowingAnswer(showingAnswer)

        // If we are in single-card mode, we show the "Show Answer" button on the question side
        // and hide navigation buttons.
        if (cardList.size == 1) {
            previewLayout!!.hideNavigationButtons()
            return
        }
        previewLayout!!.setPrevButtonEnabled(index > 0)
        previewLayout!!.setNextButtonEnabled(index < cardList.size - 1)
    }

    private fun updateProgress() {
        if (progressSeekBar.progress != index) {
            progressSeekBar.progress = index
        }
        val progress = getString(R.string.preview_progress_bar_text, index + 1, cardList.size)
        progressText.text = progress
    }

    private fun setOkResult() {
        setResult(
            RESULT_OK,
            Intent().apply {
                putExtra("reloadRequired", reloadRequired)
                putExtra("noteChanged", noteChanged)
            }
        )
    }

    companion object {
        @CheckResult
        fun PreviewDestination.toIntent(context: Context) =
            Intent(context, Previewer::class.java).apply {
                putExtra("index", index)
                putExtra("cardList", cardList)
            }
    }
}

class PreviewDestination(val index: Int, val cardList: LongArray)
